src/library/type_context.cpp

```diff
@@ -4097,22 +4098,24 @@ expr type_context_old::preprocess_class(expr const & type,
     if (!u_replacements.empty())
         C = update_constant(C, to_list(C_levels));
     expr it2 = infer(C);
+    buffer<bool> is_out_param;
+    class_out_param_deps(it2, is_out_param);
+    int i = 0;
     for (expr & C_arg : C_args) {
         it2  = whnf(it2);
         if (!is_pi(it2))
             return type; /* failed */
         expr const & d = binding_domain(it2);
-        if (/* Case 1 */
-            is_class_out_param(d) ||
-            /* Case 2 */
-            depends_on_mvar(d, new_mvars)) {
+
+        if (is_out_param[i]) {
```

Comment:
Sorry, I see another issue now.

```suggestion
        if (i < is_out_param.size() && is_out_param[i]) {
```

`class_out_param_deps` computes the dependencies literally without reducing the type, so the array might contain fewer elements than the number of arguments here.  (I don't think this is actually used anywhere though.)